File: labs/control-ui/api-shadow/stealth_bridge_collector.ts
// Control UI TypeScript: Tor Pluggable Transport Stealth Bridge Collector API

export type PluggableTransportType = 'obfs4' | 'snowflake' | 'webtunnel' | 'meek' | 'custom';

export interface StealthBridge {
  transport: PluggableTransportType;
  endpoint: string;
  fingerprint: string;
  params: Record<string, string>;
  score: number;
  latencyMs: number;
  verified: boolean;
}

export class StealthBridgeCollector {
  public minScoreThreshold: number;
  public bridges: Map<string, StealthBridge> = new Map();

  constructor(minScoreThreshold: number = 0.5) {
    this.minScoreThreshold = minScoreThreshold;
  }

  public parseBridgeLine(line: string): StealthBridge {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) {
      throw new Error('Empty or comment line');
    }

    const parts = trimmed.split(/\s+/);
    if (parts.length < 3) throw new Error('Insufficient tokens in bridge line');

    const proto = parts[0]!.toLowerCase();
    const transport: PluggableTransportType =
      proto === 'obfs4' || proto === 'snowflake' || proto === 'webtunnel' || proto === 'meek'
        ? (proto as PluggableTransportType)
        : 'custom';

    const endpoint = parts[1];
    const fingerprint = parts[2]!.toUpperCase();

    const params: Record<string, string> = {};
    for (const token of parts.slice(3)) {
      const idx = token.indexOf('=');
      if (idx !== -1) {
        params[token.slice(0, idx)] = token.slice(idx + 1);
      }
    }

    const bridge: StealthBridge = {
      transport,
      endpoint: endpoint!,
      fingerprint,
      params,
      score: 1.0,
      latencyMs: 0,
      verified: false,
    };

    this.bridges.set(fingerprint, bridge);
    return bridge;
  }

  public recordHealth(fingerprint: string, latencyMs: number, success: boolean): boolean {
    const bridge = this.bridges.get(fingerprint.toUpperCase());
    if (!bridge) return false;

    if (success) {
      bridge.verified = true;
      bridge.latencyMs = latencyMs;
      const factor = Math.min(2.0, 1000.0 / Math.max(50, latencyMs));
      bridge.score = bridge.score * 0.8 + 1.2 * factor;
    } else {
      bridge.score *= 0.5;
      if (bridge.score < 0.1) bridge.verified = false;
    }
    return true;
  }

  public getBestBridges(transport?: PluggableTransportType, limit: number = 5): StealthBridge[] {
    return Array.from(this.bridges.values())
      .filter((b) => (!transport || b.transport === transport) && b.score >= this.minScoreThreshold)
      .sort((a, b) => b.score - a.score)
      .slice(0, limit);
  }
}

File: labs/mobile/donor-archive/tunnel/StealthBridgeCollector.kt
package com.luminet.android.tunnel

import java.util.concurrent.ConcurrentHashMap

enum class AndroidPluggableTransport {
    OBFS4,
    SNOWFLAKE,
    WEBTUNNEL,
    MEEK,
    CUSTOM
}

data class AndroidStealthBridge(
    val transport: AndroidPluggableTransport,
    val endpoint: String,
    val fingerprint: String,
    val params: Map<String, String>,
    var score: Double = 1.0,
    var latencyMs: Int = 0,
    var verified: Boolean = false
)

class StealthBridgeCollector(val minScoreThreshold: Double = 0.5) {
    private val bridges = ConcurrentHashMap<String, AndroidStealthBridge>()

    fun parseBridgeLine(line: String): AndroidStealthBridge {
        val trimmed = line.trim()
        if (trimmed.isEmpty() || trimmed.startsWith("#")) {
            throw IllegalArgumentException("Empty or comment line")
        }

        val parts = trimmed.split(Regex("\\s+"))
        if (parts.size < 3) throw IllegalArgumentException("Insufficient tokens")

        val transport = when (parts[0].lowercase()) {
            "obfs4" -> AndroidPluggableTransport.OBFS4
            "snowflake" -> AndroidPluggableTransport.SNOWFLAKE
            "webtunnel" -> AndroidPluggableTransport.WEBTUNNEL
            "meek" -> AndroidPluggableTransport.MEEK
            else -> AndroidPluggableTransport.CUSTOM
        }

        val endpoint = parts[1]
        val fingerprint = parts[2].uppercase()

        val params = mutableMapOf<String, String>()
        for (token in parts.drop(3)) {
            val kv = token.split("=", limit = 2)
            if (kv.size == 2) {
                params[kv[0]] = kv[1]
            }
        }

        val bridge = AndroidStealthBridge(
            transport = transport,
            endpoint = endpoint,
            fingerprint = fingerprint,
            params = params
        )
        bridges[fingerprint] = bridge
        return bridge
    }

    fun recordHealth(fingerprint: String, latencyMs: Int, success: Boolean): Boolean {
        val bridge = bridges[fingerprint.uppercase()] ?: return false
        if (success) {
            bridge.verified = true
            bridge.latencyMs = latencyMs
            val factor = minOf(2.0, 1000.0 / maxOf(50, latencyMs))
            bridge.score = (bridge.score * 0.8) + (1.2 * factor)
        } else {
            bridge.score *= 0.5
            if (bridge.score < 0.1) bridge.verified = false
        }
        return true
    }

    fun getBestBridges(transport: AndroidPluggableTransport?, limit: Int): List<AndroidStealthBridge> {
        return bridges.values
            .filter { (transport == null || it.transport == transport) && it.score >= minScoreThreshold }
            .sortedByDescending { it.score }
            .take(limit)
    }
}
